fetch/cosmwasm_contract/src/contract.rs

```diff
@@ -775,27 +879,56 @@ fn verify_refund_swap_id<S: Storage>(id: u64, storage: &S) -> HandleResult {
  * ************    Access Control      ***************
  * ***************************************************/
 
-fn only_admin<S: Storage>(env: &Env, storage: &S) -> HandleResult {
-    _only_role(&AccessRole::Admin, env, storage)
+fn only_admin<S: Storage, A: Api>(env: &Env, storage: &S, api: &A) -> HandleResult {
+    _only_role(&AccessRole::Admin, env, storage, api)
 }
 
-fn only_relayer<S: Storage>(env: &Env, storage: &S) -> HandleResult {
-    _only_role(&AccessRole::Relayer, env, storage)
+fn only_relayer<S: Storage, A: Api>(env: &Env, storage: &S, api: &A) -> HandleResult {
+    _only_role(&AccessRole::Relayer, env, storage, api)
 }
 
-fn only_delegate<S: Storage>(env: &Env, storage: &S) -> HandleResult {
-    _only_role(&AccessRole::Delegate, env, storage)
+fn only_approver<S: Storage, A: Api>(env: &Env, storage: &S, api: &A) -> HandleResult {
+    _only_role(&AccessRole::Approver, env, storage, api)
 }
 
-fn _only_role<S: Storage>(role: &AccessRole, env: &Env, storage: &S) -> HandleResult {
-    match ac_have_role(storage, &env.message.sender, role) {
-        Ok(_) => Ok(HandleResponse::default()),
+fn only_monitor<S: Storage, A: Api>(env: &Env, storage: &S, api: &A) -> HandleResult {
+    _only_role(&AccessRole::Monitor, env, storage, api)
+}
+
+fn _only_role<S: Storage, A: Api>(
+    role: &AccessRole,
+    env: &Env,
+    storage: &S,
+    api: &A,
+) -> HandleResult {
+    let env_message_sender = api.human_address(&env.message.sender)?;
+
+    match ac_have_role(storage, &env_message_sender, role) {
+        Ok(has_role) => match has_role {
+            true => Ok(HandleResponse::default()),
+            false => Err(StdError::generic_err(match role {
+                AccessRole::Admin => ERR_ACCESS_CONTROL_ONLY_ADMIN,
+                AccessRole::Relayer => ERR_ACCESS_CONTROL_ONLY_RELAYER,
+                _ => ERR_ACCESS_CONTROL_ONLY_ADMIN,
+            })),
+        },
         Err(err) => Err(err),
     }
 }
 
-fn can_pause<S: Storage>(env: &Env, storage: &S) -> HandleResult {
-    only_relayer(env, storage).or(only_delegate(env, storage))
+fn can_pause<S: Storage, A: Api>(
+    env: &Env,
+    storage: &S,
+    api: &A,
+    since_block: u64,
+) -> HandleResult {
+    if since_block <= env.block.height {
+        // pausing
+        only_monitor(env, storage, api).or(only_admin(env, storage, api))
+    } else {
+        // unpausing
+        only_admin(env, storage, api)
+    }
```

Comment:
Just nit picking - no need to change:
```suggestion
    if since_block > env.block.height {
        // unpausing
        only_admin(env, storage, api)
    } else {
        // pausing
        only_monitor(env, storage, api).or(only_admin(env, storage, api))
    }
```